### ISSUE TITLE
Unable to execute dex: Multiple dex files define Lorg/apache/http/entity/mime/FormBodyPart

### ISSUE THREAD

**Issue by pardypanda** [QuickBlox/quickblox-android-sdk#61]
Hi, 

I'm trying to implement QB chat in my already existing app.

My project build goes through smoothly, however, when I try to run the project as an Android application, I get the following error:

[2015-01-15 16:03:31 - Dex Loader] Unable to execute dex: Multiple dex files define Lorg/apache/http/entity/mime/FormBodyPart;
[2015-01-15 16:03:31 - MyApp] Conversion to Dalvik format failed: Unable to execute dex: Multiple dex files define Lorg/apache/http/entity/mime/FormBodyPart;

I have multiple other jars included in my project and I believe it is conflicting with one of them. A screenshot of all the jars in my project is here: http://imgur.com/gBZJ7Mw

I also tried removing the httpmime.jar but that starts giving me compile time errors as that jar is required for using multipartrequest functions for image upload. I think QB is also using multipartrequest, but via it's own methods and here is where the above conflict is arising.

Now, I need to find a solution where my existing code also runs and I can easily incorporate QB-Core and QB-Chat jars in my project and be able to run the project as an Android app.

Please help, it's urgent.

Thanks

**Comment by soulfly** [comment]
Yes, Android SDK uses httpmime.jar

try to remove this jar from your app and use the embedded one from QB sdk

**Comment by pardypanda** [comment]
Ok, one solution to conflicting jars I found online was using jarjar tool to repackage the jar to a different package name.

I tried repackaging the QB-core and QB-chat jars like this:
java -jar jarjar-1.4.jar process rules.txt quickblox-android-sdk-core-2.0.jar qb-quickblox-android-sdk-core-2.0.jar 
java -jar jarjar-1.4.jar process rules.txt quickblox-android-sdk-chat-2.0.jar qb-quickblox-android-sdk-chat-2.0.jar 


with rules.txt:
rule org.apache.http.** QB.org.apache.http.@1


So, now after that, there are no build errors or compile time errors, and the app started on the device. However, now I'm getting NoClassDefFound errors. Stack trace:

01-16 12:11:28.592: E/AndroidRuntime(4437): java.lang.NoClassDefFoundError: Failed resolution of: LQB/org/apache/http/client/methods/HttpPost;
01-16 12:11:28.592: E/AndroidRuntime(4437): 	at com.quickblox.core.rest.RestRequest.getHttpRequest(RestRequest.java:128)
01-16 12:11:28.592: E/AndroidRuntime(4437): 	at com.quickblox.core.rest.RestRequest.asyncRequestWithCallback(RestRequest.java:100)
01-16 12:11:28.592: E/AndroidRuntime(4437): 	at com.quickblox.core.query.Query.performInBgAsyncWithDelegate(Query.java:114)
01-16 12:11:28.592: E/AndroidRuntime(4437): 	at com.quickblox.core.query.Query.performAsyncWithCallback(Query.java:216)
01-16 12:11:28.592: E/AndroidRuntime(4437): 	at com.quickblox.auth.QBAuth.createSession(QBAuth.java:176)
01-16 12:11:28.592: E/AndroidRuntime(4437): 	at com.pardypanda.play.ChatTest.onCreate(ChatTest.java:51)
01-16 12:11:28.592: E/AndroidRuntime(4437): 	at android.app.Activity.performCreate(Activity.java:5933)
01-16 12:11:28.592: E/AndroidRuntime(4437): 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1105)
01-16 12:11:28.592: E/AndroidRuntime(4437): 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2251)
01-16 12:11:28.592: E/AndroidRuntime(4437): 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2360)
01-16 12:11:28.592: E/AndroidRuntime(4437): 	at android.app.ActivityThread.access$800(ActivityThread.java:144)
01-16 12:11:28.592: E/AndroidRuntime(4437): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1278)
01-16 12:11:28.592: E/AndroidRuntime(4437): 	at android.os.Handler.dispatchMessage(Handler.java:102)
01-16 12:11:28.592: E/AndroidRuntime(4437): 	at android.os.Looper.loop(Looper.java:135)
01-16 12:11:28.592: E/AndroidRuntime(4437): 	at android.app.ActivityThread.main(ActivityThread.java:5221)
01-16 12:11:28.592: E/AndroidRuntime(4437): 	at java.lang.reflect.Method.invoke(Native Method)
01-16 12:11:28.592: E/AndroidRuntime(4437): 	at java.lang.reflect.Method.invoke(Method.java:372)
01-16 12:11:28.592: E/AndroidRuntime(4437): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:899)
01-16 12:11:28.592: E/AndroidRuntime(4437): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:694)
01-16 12:11:28.592: E/AndroidRuntime(4437): Caused by: java.lang.ClassNotFoundException: Didn't find class "QB.org.apache.http.client.methods.HttpPost" on path: DexPathList[[zip file "/data/app/com.pardypanda.play-1/base.apk"],nativeLibraryDirectories=[/vendor/lib, /system/lib]]
01-16 12:11:28.592: E/AndroidRuntime(4437): 	at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
01-16 12:11:28.592: E/AndroidRuntime(4437): 	at java.lang.ClassLoader.loadClass(ClassLoader.java:511)
01-16 12:11:28.592: E/AndroidRuntime(4437): 	at java.lang.ClassLoader.loadClass(ClassLoader.java:469)
01-16 12:11:28.592: E/AndroidRuntime(4437): 	... 19 more
01-16 12:11:28.592: E/AndroidRuntime(4437): 	Suppressed: java.lang.ClassNotFoundException: QB.org.apache.http.client.methods.HttpPost
01-16 12:11:28.592: E/AndroidRuntime(4437): 		at java.lang.Class.classForName(Native Method)
01-16 12:11:28.592: E/AndroidRuntime(4437): 		at java.lang.BootClassLoader.findClass(ClassLoader.java:781)
01-16 12:11:28.592: E/AndroidRuntime(4437): 		at java.lang.BootClassLoader.loadClass(ClassLoader.java:841)
01-16 12:11:28.592: E/AndroidRuntime(4437): 		at java.lang.ClassLoader.loadClass(ClassLoader.java:504)
01-16 12:11:28.592: E/AndroidRuntime(4437): 		... 20 more
01-16 12:11:28.592: E/AndroidRuntime(4437): 	Caused by: java.lang.NoClassDefFoundError: Class not found using the boot class loader; no stack available


I unzipped and opened up the original QB-Chat and QB-Core jars, and did not find any org/apache/http/client/ (or QB/org/apache/http/client/ in the repackaged folder) folders. I only find org/apache/http/mime (or QB/org/apache/http/mime/ in the repackaged folder) folder. Is this how it's supposed to be? So, I don't know if this is a repackaging issue or not? It doesn't seem like it. 


Any idea how we can fix the http/client etc classes issue?

**Comment by soulfly** [comment]
Did you try my recommendation?

**Comment by pardypanda** [comment]
Yes,

But I'm using httpmime-4.3.5 and there a lot of multipart type classes that were not found from the Qb-Chat or QB-Core jars and were giving compile errors. So, I tried the above method.

**Comment by nbayaman** [comment]
Hello,

I'm getting also following error:
```
  	UNEXPECTED TOP-LEVEL EXCEPTION:
  	com.android.dex.DexException: Multiple dex files define Lorg/apache/http/entity/mime/FormBodyPart;
```
Which is also caused by httpmime library import duplication. Would you consider including quickblox libraries as a standalone ones, without wrapping inside other libraries? It would help to manage needed libraries. As it's been said already - included httpmime library does not include some essential classes, Namely MultipartFormBuilder.

**Comment by soulfly** [comment]
Hi guys, 

do you have the test demo Android app where we can face this error?

**Comment by nbayaman** [comment]
Actually with this conflict we could not compile a working app. We solved this problem by pulling needed code out of httpmime library and use quickblox sdk libraries. But again, our suggestion is about logic of library packaging. Wouldn't it be more logic and clean to keep external libraries outside of your libraries?

**Comment by ibnemuhammad** [comment]
Is there no solution found for this?

**Comment by soulfly** [comment]
resolved in 2.5 SDK

**Comment by qydq** [comment]
dear ,i'm getting same error ,you can look at this 

"Error:Execution failed for task ':app:transformClassesWithDexForDebug'.
> com.android.build.api.transform.TransformException: com.android.ide.common.process.ProcessException: java.util.concurrent.ExecutionException: com.android.ide.common.process.ProcessException: org.gradle.process.internal.ExecException: Process 'command 'D:\ide\jdk1.8.0_102\bin\java.exe'' finished with non-zero exit value 2"

what is httpmime.jar ? ， someone said we can use jdk1.7 , i try ,but it is not work ,the same project can run with my friend ,his android studio version is 1.3 (it's too old ) andｉ　update　ｆｏｒ　ｓｔｕｄｉｏ　ｉｓ　２．２　－　ｐｒｅｖｉｅｗ　６　，ｙｏｕ　ｃａｎ　ｓｅｅ　ｍｙ　ｇｒａｄｌｅ．ｂｕｉｄ　ｆｏｒ　ａｐｐ　ｐｒｏｐｅｒｔｉｅｓ．
dependencies {
    compile fileTree(include: ['*.jar'], dir: 'libs')
    compile files('libs/android-async-http-1.4.5.jar')
    compile 'com.android.support:appcompat-v7:23.2.0'
    compile 'com.android.support:design:23.2.0'
    compile 'com.android.support:support-v4:23.2.0'
    compile 'me.drakeet.materialdialog:library:1.2.8'
    compile 'com.pnikosis:materialish-progress:1.7'
    compile files('src/main/libs/nineoldandroids-2.4.0.jar')
    compile files('libs/gson-2.2.2.jar')
    compile files('libs/xUtils-2.6.14.jar')
    compile files('libs/org.apache.http.legacy.jar')
    compile files('libs/hirondelle-date4j-1.5.0.jar')
    compile files('libs/zxing.jar')

ａｎｄ　ｙｏｕ　ｃａｎ　ｌｏｏｋ　ｕｐ　ｐｒｏｊｅｃｔ　ｇｒａｄｌｅ．ｂｕｉｌｄ　　ｐｒｏｐｅｒｔｉｅｓ　：
dependencies {
//        classpath 'com.android.tools.build:gradle:1.3.0'
//        classpath 'com.android.tools.build:gradle:2.2.+'
//        classpath 'com.android.tools.build:gradle:2.1.2'
        classpath 'com.android.tools.build:gradle:2.1.0'
//        classpath 'com.android.tools.build:gradle:2.0.0'
//        classpath 'com.android.tools.build:gradle:1.5.0'
//        classpath 'com.android.tools.build:gradle:2.2.0-alpha6'

        // NOTE: Do not place your application dependencies here; they belong
        // in the individual module build.gradle files
    }

ｔｈｉｓ　ｉｓ　ｄｉｆｆｉｃｕｌｔ　；ａｎｙｏｎｅ　ｓｏｌｕｔｉｏｎ　ｔｈｅ　ｐｒｏｂｌｅｍ？
